iexec-sms-library/src/main/java/com/iexec/sms/api/SmsClientProvider.java

```diff
@@ -0,0 +1,103 @@
+package com.iexec.sms.api;
+
+import com.iexec.common.chain.ChainDeal;
+import com.iexec.common.task.TaskDescription;
+import com.iexec.common.lifecycle.purge.ExpiringTaskMapFactory;
+import com.iexec.common.lifecycle.purge.Purgeable;
+import feign.Logger;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Manages the {@link SmsClient}, providing an easy way to access SMS
+ * and avoiding the need to create a new {@link SmsClient} instance each time.
+ */
+@Slf4j
+public class SmsClientProvider implements Purgeable {
+    private final Map<String, Optional<String>> taskIdToSmsUrl =
+            ExpiringTaskMapFactory.getExpiringTaskMap();
+    private final Map<String, SmsClient> urlToSmsClient = new HashMap<>();
+
+    private final Logger.Level loggerLevel;
+
+    public SmsClientProvider() {
+        this(Logger.Level.BASIC);
+    }
+
+    public SmsClientProvider(Logger.Level loggerLevel) {
+        this.loggerLevel = loggerLevel;
+    }
+
+    /**
+     * Retrieves the specified SMS URL for this task, then:
+     * <ul>
+     *     <li>If this SMS has already been accessed, returns the already-constructed {@link SmsClient};</li>
+     *     <li>Otherwise, constructs, stores and returns a new {@link SmsClient}.</li>
+     * </ul>
+     *
+     * @param taskDescription Task the specified SMS URL should be retrieved for.
+     * @throws SmsClientCreationException if SMS URL can't be retrieved.
+     * @return An instance of {@link SmsClient} pointing on the task's specified SMS.
+     */
+    public SmsClient getOrCreateSmsClientForTask(TaskDescription taskDescription) {
+        final String chainTaskId = taskDescription.getChainTaskId();
+
+        final Optional<String> smsUrl = taskIdToSmsUrl.computeIfAbsent(
```

Comment:
With a second reading, I'm not sure this wrapping in `Optional` is really necessary here.
The data is wrapped once and can be unwrapped twice without really avoiding a `null` check.
Do I miss something ?

Reply:
You're right. That's probably an old implementation that has not been cleaned when moving to a new state. FIXME added in https://github.com/iExecBlockchainComputing/iexec-sms/pull/117/commits/e966d64155c9e1269d82f8b5679fb9f536f5f734.